src/index.ts

```diff
@@ -0,0 +1,83 @@
+import { bytes_to_hex, Sha256 } from "asmcrypto.js"
+import {
+  TFileMetadata,
+  TdrFile,
+  TGenerateMetadata,
+  TTotalChunks,
+  TSliceToArray,
+  TChunkProgressFunction
+} from "./types"
+
+export const extractFileMetadata: TFileMetadata = async (
+  files,
+  progressFunction,
+  chunkSize = 100000000
+) => {
+  let processedChunks = 0
+  const totalChunks = getTotalChunks(files, chunkSize)
+
+  const chunkProgress: TChunkProgressFunction = () => {
+    processedChunks += 1
+    const ratioProcessed = processedChunks / totalChunks
+    const percentageProcessed = Math.round(ratioProcessed * 100)
+    const totalFiles = files.length
+    const processedFiles = Math.floor(ratioProcessed * totalFiles)
```

Comment:
This number's a bit misleading because it's a kind of average. Can we keep track of the actual number of processed files?

This is why I was a bit apprehensive about making this a library from the start - we're sort of guessing about what we need. :slightly_smiling_face:

Reply:
It's not average, `Math.floor` makes it exactly right. It is quite a lot more difficult to pass around the number of processed files variables and as this produces exactly the same result,

Reply:
I've just had a think about this actually and I think you're right, let me check it.

Reply:
I think the problem with this is that because getting the byte array from the file is async, the percentage progress, i.e. how many chunks have been processed and how many files have been processed, i.e. how often the generateHash function completes, are not always going to be in sync because the function can complete before every await inside the loop has finished processing. If I try passing the number of files variable in, it doesn't always match up with the percentage because of this. 
In short, I don't think we can have both and I think percentage is more useful because it will still increment regularly with large files so I'll just remove the file count.
If we really want a file count, we can just pass files in individually and count them on the front end.
I've taken out the file count and because the total number of files doesn't make much sense without it, I've taken that out as well.

Reply:
Actually, the above explanation might not be totally accurate but there's definitely some async strangeness going on here that's preventing accurate reporting of which file we're up to for a particular percentage and as it's arguably not that valuable, I still think we should just take it out.

Reply:
Ok, I think I was right although I'm not exactly sure why. I've removed the call to `files.map` and put in a good old fashioned for loop with another one for the chunks in the same function and this is now giving sensible results and it's probably easier to read as well.

The point still stands about whether we want to have a processed files count but now I've managed to get it working, I've changed my mind about it :smile:

Reply:
Cheers for looking into it! I haven't followed all the steps you've taken, but the current version looks like it gives consistent results so let's try it for now. 👍 We can come back to it if it looks like it might be more useful to have an overall percentage, or something else.